### ISSUE TITLE
Explicit provider: assumeRole and/or roleArn are set, but no role is assumed

### ISSUE THREAD

**Issue by veikak** [pulumi/pulumi-aws-native#1279]
### What happened?

I'm using explicit AWS Native providers in TypeScript. I'm trying to do some bog-standard role chaining stuff, but I can't get role assumption to work with the provider. I have multiple AWS accounts, so the providers can't all use the same role. That is, I can't just assume a single role externally and let Pulumi use that. Locally, everything works just fine with SSO profiles (using the `profile` property on the provider). In our CI I'd like to pass each provider a role ARN to assume instead, but I just can't get it to work.

This might be irrelevant, but my setup is as follows:
- GitHub Actions authenticates to a bastion AWS account with OIDC, and assumes `arn:aws:iam::<bastion-account-id>:role/bastion-role`.
- The AWS Native providers read the OIDC credentials.
- The providers are configured to assume `arn:aws:iam::<target-account-id>:role/target-role` in another account. Each target account has a copy of this role.
- The roles have appropriate permissions and trust relationships.
  - `bastion-role` is allowed to assume all of the `target-role`s.
  - The `target-role`s trust `bastion-role`, and they have the permissions Pulumi needs.
  - I've confirmed that the roles work as expected. Pulumi can execute a `preview --refresh` if I assume the role externally and pass the credentials to the Pulumi program.

#### What I tried

Run `pulumi preview --refresh` against any of my stacks in GitHub Actions.

#### Expected outcome

- The provider should assume `target-role` with GHA's OIDC credentials, and use it for all CloudControl operations.
- The AWS account id the provider resolves should be the target account's id.

#### Actual outcome

```
error: Preview failed: operation error CloudControl: GetResource, https response error StatusCode: 400, RequestID: <snip>, api error AccessDeniedException: User: arn:aws:sts::<bastion-account-id>:assumed-role/bastion-role/GitHubActions is not authorized to perform: cloudformation:GetResource on resource: arn:aws:cloudformation:eu-central-1:<bastion-account-id>:resource/* because no identity-based policy allows the cloudformation:GetResource action
```
That is, the original role assumed by GHA is used for the CloudControl calls instead of the role I want. Additionally, the AWS account id the provider resolves is the bastion account's id (`allowedAccountIds` validation fails if it's set to `["<target-account-id>"]`).

#### TL;DR

I'm telling AWS Native to assume a (chained) role, but it doesn't want to do it. There's no indication whatsoever that I did something wrong, or that something might have gone wrong with the role assumption. The provider just happily uses the original role and targets the wrong AWS account.

### Example

Again, the exact setup probably isn't relevant, but here it is anyway.

Below, there are two programs that'll bootstrap a minimal IAM setup, and a program that demonstrates how role assumption doesn't work.

Run this against the bastion account:
```ts
import * as aws from "@pulumi/aws-native"

const awsAccountIds = {
  Bastion: "<bastion-account-id>",
  Target: "<target-account-id>"
}

const oidcProvider = new aws.iam.OidcProvider("github-oidc-provider", {
  url: "https://token.actions.githubusercontent.com",
  clientIdList: ["sts.amazonaws.com"],
  // Unused, but required: https://github.com/aws-actions/configure-aws-credentials#sample-iam-oidc-cloudformation-template
  thumbprintList: ["ffffffffffffffffffffffffffffffffffffffff"]
})

const bastionRole = new aws.iam.Role("bastion-role", {
  // We need a fixed name to use this role in an external service
  roleName: "bastion-role",
  assumeRolePolicyDocument: {
    Version: "2012-10-17",
    Statement: [
      {
        Effect: "Allow",
        Principal: {
          Federated: oidcProvider.arn
        },
        Action: "sts:AssumeRoleWithWebIdentity",
        Condition: {
          StringLike: {
            "token.actions.githubusercontent.com:sub": "repo:<org>/<repo>:*"
          },
          StringEquals: {
            "token.actions.githubusercontent.com:aud": "sts.amazonaws.com"
          }
        }
      }
    ]
  },
  policies: [
    {
      policyName: `AllowAssumeCiRoles`,
      policyDocument: JSON.stringify({
        Version: "2012-10-17",
        Statement: Object.entries(awsAccountIds)
          .sort(([a], [b]) => a.localeCompare(b))
          .map(([policyAccountName, policyAccountId]) => ({
            Sid: `Allow${policyAccountName}`,
            Effect: "Allow",
            Resource: `arn:aws:iam::${policyAccountId}:role/target-role`,
            Action: "sts:AssumeRole"
          }))
      })
    }
  ]
})

const targetRole = new aws.iam.Role("target-role", {
  roleName: "target-role",
  assumeRolePolicyDocument: {
    Version: "2012-10-17",
    Statement: [
      {
        Effect: "Allow",
        Principal: {
          AWS: bastionRole.arn
        },
        Action: "sts:AssumeRole"
      }
    ]
  },
  managedPolicyArns: ["arn:aws:iam::aws:policy/SecurityAudit"],
  policies: [
    {
      policyName: `AllowWhatSecurityAuditIsMissing`,
      policyDocument: JSON.stringify({
        Version: "2012-10-17",
        Statement: [
          {
            Effect: "Allow",
            Resource: "arn:aws:cloudformation:eu-central-1:*:resource/*",
            Action: "cloudformation:*"
          }
        ]
      })
    }
  ]
})
```

Then run this against the target account:
```ts
import * as aws from "@pulumi/aws-native"

const awsAccountIds = {
  Bastion: "<bastion-account-id>",
  Target: "<target-account-id>"
}

const targetRole = new aws.iam.Role("target-role", {
  roleName: "target-role",
  assumeRolePolicyDocument: {
    Version: "2012-10-17",
    Statement: [
      {
        Effect: "Allow",
        Principal: {
          // Let's cheat a bit to avoid stack references
          AWS: `arn:aws:iam::${awsAccountIds.Bastion}:role/bastion-role`
        },
        Action: "sts:AssumeRole"
      }
    ]
  },
  managedPolicyArns: ["arn:aws:iam::aws:policy/SecurityAudit"],
  policies: [
    {
      policyName: `AllowWhatSecurityAuditIsMissing`,
      policyDocument: JSON.stringify({
        Version: "2012-10-17",
        Statement: [
          {
            Effect: "Allow",
            Resource: "arn:aws:cloudformation:eu-central-1:*:resource/*",
            Action: "cloudformation:*"
          }
        ]
      })
    }
  ]
})
```

Run both of these and you'll have a `bastion-role` in the bastion account, and a `target-role` in both accounts. If you assume `bastion-role` with OIDC, you should be able to assume either `target-role`.

Now, we can try to assume a role in another program. Running this against either account should work (change `targetAccountName` accordingly), but doesn't. A role isn't assumed.
```ts
import * as aws from "@pulumi/aws-native"

const awsAccountIds = {
  Bastion: "<bastion-account-id>",
  Target: "<target-account-id>"
}

const targetAccountName: keyof typeof awsAccountIds = "Target"

const provider = new aws.Provider("test-provider", {
  region: "eu-central-1",
  accessKey: process.env.AWS_ACCESS_KEY_ID,
  secretKey: process.env.AWS_SECRET_ACCESS_KEY,
  token: process.env.AWS_SESSION_TOKEN,
  assumeRole: {
    // Let's cheat a bit to avoid stack references
    roleArn: `arn:aws:iam::${awsAccountIds[targetAccountName]}:role/target-role`,
    sessionName: `pulumi.test${
      process.env.GITHUB_RUN_ID !== undefined
        ? `.gha-run-${process.env.GITHUB_RUN_ID}`
        : ""
    }`
  },
  sharedCredentialsFile: "",
  skipCredentialsValidation: false,
  skipGetEc2Platforms: false,
  skipMetadataApiCheck: false,
  skipRegionValidation: false
})

const testGroup = new aws.iam.Group(
  "test-group",
  {},
  {
    provider
  }
)
```

### Output of `pulumi about`

I have a multi-project setup with shared dependencies, and the NPM package belongs to a Yarn workspace. The about command doesn't seem be aware of any of that.
```
$ pulumi -C projects/test about
CLI
Version      3.101.1
Go Version   go1.21.5
Go Compiler  gc

Plugins
NAME    VERSION
nodejs  unknown

Host
OS       darwin
Version  13.6.1
Arch     arm64

This project is written in nodejs: executable='/Users/<snip>/.nvm/versions/node/v18.14.2/bin/node' version='v18.14.2'

Current Stack: <snip>/test/data-science

TYPE                         URN
pulumi:pulumi:Stack          urn:pulumi:data-science::test::pulumi:pulumi:Stack::test-data-science
pulumi:providers:aws-native  urn:pulumi:data-science::test::pulumi:providers:aws-native::test-provider
aws-native:iam:Group         urn:pulumi:data-science::test::aws-native:iam:Group::test-group


Found no pending operations associated with data-science

Backend
Name           pulumi.com
URL            https://app.pulumi.com/<snip>
User           <snip>
Organizations  <snip>
Token type     personal

Pulumi locates its logs in /var/folders/s2/dxll8hdj7xs4rt4ny2p47f640000gp/T/ by default
warning: Failed to get information about the Pulumi program's dependencies: could not find either <snip>/devops/pulumi/projects/test/yarn.lock or <snip>/devops/pulumi/projects/test/package-lock.json
```

Here's the package.json because Pulumi couldn't find the info on its own:
```json
{
  "name": "@<snip>/pulumi",
  "version": "0.0.0-internal",
  "private": true,
  "scripts": {
    "snip": "snip"
  },
  "devDependencies": {
    "@types/node": "^18",
    "@types/ramda": "^0"
  },
  "dependencies": {
    "@pulumi/aws-native": "0.93.0",
    "@pulumi/pulumi": "^3.0.0",
    "luxon": "1.28.0",
    "ramda": "^0.29.1",
    "ts-node": "^10.9.2"
  }
}
```

### Additional context

I tried both `assumeRole` and `roleArn`, separately and together. Neither of them seemed to have any effect on anything. Max verbosity and debug logs didn't seem to contain anything helpful.

The bastion account itself has a `target-role` too, but it doesn't work either.

I hope I'm not missing something obvious here 😄 I could work around this by assuming the roles externally and marshaling the creds to the programs, but I'd rather not introduce more complexity when the provider should be able to do this on its own.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by veikak** [comment]
I figured it out... I was under the impression the `ignoreChanges` option would make the provider skip writing the props to the state, but still use the supplied values. Of course that's not the case. I was using ignores to prevent unnecessary changes to the profile/role outputs, which made the provider stick to whatever props the state had before I set up the ignores 🤦‍♂️

**Comment by mjeffryes** [comment]
Thanks for reporting this issue @veikak. We do have an example for role assumption https://github.com/pulumi/pulumi-aws-native/tree/2e477cc7d1dbc84386adba285e6ceb347b5ef254/examples/aws-ts-assume-role, but I don't think it's run under test so we may have regressed. Probably a first debug step here will be to check if that example exhibits the same problem.

**Comment by veikak** [comment]
@mjeffryes Thanks. I walked through the example, and it worked just fine. I also modified it a bit to try role chaining (unprivileged-user -> new jump role I added -> allow-loggroup-management) and it worked too.

But I think I figured this out. I think `pulumi preview --refresh` uses the current state of the provider in the refresh phase. That is, even the authentication props come directly from the state. In my case, I deployed the stacks from my computer so `profile` is set and `assumeRole` is not. The CI setup I have runs the update with `profile` unset and `assumeRole` set, but the new values aren't used in the refresh phase.

To reproduce:

1. Walk through the assume role example, but don't destroy the stacks in the end.
3. In `assume-role/index.ts`, comment out the `assumeRole` prop of the provider.
4. Run `pulumi up` (and confirm the update). You should get `[diff: -assumeRole]` without any errors.
5. Uncomment the `assumeRole` prop.
6. Run `pulumi preview`. You should get `[diff: +assumeRole]` without any errors.
7. Run `pulumi preview --refresh`.
    - My expectation: You get `[diff: +assumeRole]` without any errors.
    - Reality: You get a CloudControl error.

I have no idea whether this is a bug or by design. And I don't really like the fact that the provider stores it's auth info in the state because I need to change the `profile` and `assumeRole` props depending on where I run Pulumi 👀 This makes `--expect-no-changes` practically useless for me, and I get needless spam from Pulumi's GH bot because there'll always be trivial changes in every stack.

**Comment by mjeffryes** [comment]
Ah, thank you for digging in deeper here. Indeed, a refresh is designed to run completely independently of the program, so it uses the provider config from the state file. I can see how that totally does not match expectations here.

As a partial workaround, you could only update state (ie. use `pulumi refresh`, `pulumi preview --refresh` or `pulumi update`) from CI? That would keep your state in the correct condition for running `pulumi preview --refresh` from CI. You'd still see spurious diffs for `assumeRole` running locally.

**Comment by veikak** [comment]
> As a partial workaround, you could only update state (ie. use pulumi refresh, pulumi preview --refresh or pulumi update) from CI?

Ideally, yes, but I'm sure we'll need to do updates from dev computers occasionally. Any off-CI updates would break previews in CI until the next update from CI.

I managed to work around the authentication issue by acquiring the creds externally and using the provider's credential props (`accessKey` etc.) directly. This works perfectly, but leaves me with a cred prop diff on every single update and preview because the creds change every time.

**Comment by danielrbradley** [comment]
I believe this is being tracked by:
- https://github.com/pulumi/pulumi/issues/13860

This behaviour is currently a core part of pulumi as @mjeffryes mentions because the program is not currently executed during a refresh - so the configuration can't be updated either.

Another workaround for this kind of issue is to use environment variables for authentication configuration so it doesn't get persisted into your state and it can be updated independently to the program.

I'm going to close this issue as a duplicate of https://github.com/pulumi/pulumi/issues/13860 though do let me know if this is in error.
